src/Tools/dotnet-monitor/ConfigurationJsonWriter.cs

```diff
@@ -202,28 +207,135 @@ private void ProcessSection(IConfigurationSection section, bool includeChildSect
-                    }
-                    else
+                    WriteValue(section.Value, redact);
+                }
+            }
+        }
+
+        private void ProcessChildren(IEnumerable<IConfigurationSection> children, bool includeChildSections, bool redact, IConfigurationSection section, Type sectionType)
+        {
+            foreach (IConfigurationSection child in children)
+            {
+                Type childType = GetChildType(sectionType, child.Key, section, children);
+
+                ProcessSection(child, includeChildSections, redact, childrenAreListIndices: IsListType(childType) || IsArrayType(childType), sectionType: childType);
+            }
+        }
+
+        bool IsListType(Type type)
+        {
+            return null != type && type.Name.Equals(_listName);
+        }
+
+        bool IsArrayType(Type type)
+        {
+            return null != type && type.IsArray;
+        }
+
+        private void WriteValue(string value, bool redact)
+        {
+            string valueToWrite = redact ? Strings.Placeholder_Redacted : value;
+
+            _writer.WriteStringValue(valueToWrite);
+        }
+
+        private Type GetChildType(Type sectionType, string childName, IConfigurationSection section, IEnumerable<IConfigurationSection> children)
+        {
+            if (null == sectionType)
+            {
+                return null;
+            }
+
+            Type[] arguments = sectionType.GetGenericArguments();
+
+            if (sectionType.Name.Equals(_dictionaryName))
+            {
+                return arguments[1];
+            }
+            else if (IsListType(sectionType))
+            {
+                return arguments[0];
+            }
+            else if (IsArrayType(sectionType))
+            {
+                return sectionType.GetElementType();
+            }
+            else
+            {
+                return GetMatchedPropertyType(sectionType, childName, section, children);
+            }
+        }
+
+        private Type GetMatchedPropertyType(Type sectionType, string childName, IConfigurationSection section, IEnumerable<IConfigurationSection> children)
+        {
+            var matchedProperties = sectionType.GetProperties().Where(prop => prop.Name.Equals(childName));
+
+            var matchedPropertyType = (matchedProperties.Any()) ? matchedProperties.First().PropertyType : null;
+
+            if (matchedPropertyType == typeof(object))
+            {
+                foreach (var child in children)
+                {
+                    // Special case due to not knowing the types of actions/triggers; takes advantage of the fact that we know that a Type variable must be set.
+                    if (child.Key.Equals("Type"))
                     {
-                        _writer.WriteStringValue(section.Value);
+                        string optionsType = "";
+
+                        if (section.Key.Equals(nameof(CollectionRuleOptions.Trigger)))
```

Comment:
I don't love having to special case the trigger and actions, but seeing as the design of collection rules is relatively stable I don't know that it'll really be an issue.